examples/cluster/plot_optimal_n_clusters.py

```diff
@@ -0,0 +1,73 @@
+"""Plot the results of the gap criterium."""
+
+# Authors: Thierry Guillemot <thierry.guillemot.work@gmail.com>
+
+import time
+import numpy as np
+import matplotlib.pyplot as plt
+
+from sklearn.cluster import KMeans, OptimalNClusterSearch
+from sklearn.datasets import make_blobs
+from sklearn.metrics import calinski_harabaz_score, fowlkes_mallows_score
+from sklearn.metrics import silhouette_score
+from sklearn.utils import check_random_state
+
+
+n_samples, n_features, random_state = 1000, 2, 1
+parameters = {'n_clusters': np.arange(1, 7)}
+
+rng = check_random_state(random_state)
+datasets = [
+    ('3 clusters', make_blobs(n_samples=n_samples, n_features=2,
+                              random_state=random_state, centers=3)),
+    ('5 clusters', make_blobs(n_samples=n_samples, n_features=2,
+                              random_state=random_state, centers=5)),
+    ('random', (rng.rand(n_samples, n_features),
+                np.zeros(n_samples, dtype=int))),
+]
+
+estimator = KMeans(n_init=10, random_state=0)
+searchers = [
+    ('Silhouette', OptimalNClusterSearch(
+        estimator=estimator, parameters=parameters,
+        fitting_process='unsupervised', metric=silhouette_score)),
+    ('Calinski', OptimalNClusterSearch(
+        estimator=estimator, parameters=parameters,
+        fitting_process='unsupervised', metric=calinski_harabaz_score)),
+    ('Stability', OptimalNClusterSearch(
+        estimator=estimator, parameters=parameters, random_state=0,
+        fitting_process='stability', metric=fowlkes_mallows_score)),
+    ('Distortion jump', OptimalNClusterSearch(
+        estimator=estimator, parameters=parameters,
+        fitting_process='distortion_jump')),
```

Comment:
For `unsupervised` case, the method compute the max value of an unsupervised metric passed in parameters (e.g. calinski and silhouette).
For `stability` case, the method computes the fitted labels of two subset of X and check we obtain the same clusters by applying a supervised metric (e.g. fowlkes and Mallows).
If you want I can create two new fitting process called `silhouette`, `calinski`, ... and remove the metric attribute.